tools/update.php

```diff
@@ -402,6 +460,92 @@ function applyPatches($patches, $db_config, $apply_patches = false) : bool
     return true;
 }
 
+function getPrivileges($db_config): array {
+    $A = $db_config['database'];
+    $u = $db_config['username'];
+    $h = $db_config['host'];
+    $p = $db_config['password'];
+    /* Run 'SHOW privileges' for this user and parse the output. We are passing
+     * the password value on bhelaf of the user here as the SHOW command does
+     * not alter any data.
+     *
+     * NOTE this may have issues when dealing with passwords that contain
+     * special characters that the shell may interpret differently. TBD.
+     */
+    $cmd = "echo 'show privileges' "
+        . "| mysql -h $h -u $u --password='$p' -A $A "
+        . "| cut -f 1";
+    return array_map('strtolower',
+        explode(
+            "\n", 
+            /* Don't use doExec() here because we are entering the password and 
+             * don't want it printed to stdout.
+             */
+            shell_exec($cmd)
+        )
+    );
+}
+
+
+/**
+ * Using the keywords from a MySQL script, determine whether the user in
+ * $db_config is able to execute all the commands corresponding to the keywords.
+ */
+function canApplyPatch(string $patch, $db_config): bool
+{
+    $num_missing = count(
+        getMissingPrivileges(
+            $patch,
+            $db_config
+        )
+    );
+    return $num_missing === 0;
+}
+
+/**
+ * Use bash commands to extract MySQL keywords from an sql script.
+ */
+function getKeywords(string $patch): array
+{
+    //Scan file for MySQL keywords. Exclude comments and parentheses using grep.
+    $cmd = "cut $patch -d ' ' -f 1 "
+        . "| sort -u "
+        . "| grep -v '\--' | grep -v ')'";
```

Comment:
also need to remove empty line that cut will generate on multi line instruction like CREATE TABLE.

Can use "uniq" to suppress multiple occurrences of an instruction after sort

Reply:
Empty lines are dealt with just below this by the `array_filter` call.

`sort -u` is pretty much equivalent to `sort | uniq`.